### PR TITLE
feat: Attach view hierarchy to events

### LINKED ISSUE #1772: Attach UIKit view hierarchy to events
## Description

Capture the view hierarchy and attach it to events. Can be sent as a text file via an [attachment](https://docs.sentry.io/platforms/android/enriching-events/attachments/).

Related to https://github.com/getsentry/sentry-cocoa/issues/1560 and https://github.com/getsentry/sentry-java/issues/1647

## Solution brainstorm

https://github.com/indragiek/InAppViewDebugger does something similar, but according to @indragiek 

> Hey, if you're not interested in capturing the view as an image and only need the data structure then there's no real magic to it besides iterating the view tree recursively starting at some root view and collecting the metadata attributes you're looking for (probably stuff like class name, pointer address, etc.). If you want to capture what's on screen at any given moment you'd do something like this to figure out which window is key: https://stackoverflow.com/a/57169802. And then traverse the tree starting at that window (which inherits from UIView).

### PR DESCRIPTION
**Pull request by kevinrenskers**
## :scroll: Description

All errors and exceptions can now get the full window's view hierarchy added as attachments. This is controlled by a new option `attachViewHierarchy` which I defaulted to false.

Screen Shot 2022-08-08 at 13 25 33

Screen Shot 2022-08-08 at 13 26 09

## :bulb: Motivation and Context

Closes #1772

## :green_heart: How did you test it?

Manual, and unit tests.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the docs if needed
- [x] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

**Comment by kevinrenskers** [comment]
Two questions:

1. Do I need to update any docs? If so, where?
2. Should we only add the view hierarchy attachment(s) in case of error or exception, just like the screenshots? Or with all events, as I have done now? #1772 suggests all events.

**Comment by brustolin** [comment]
> Two questions:
> 
> 1. Do I need to update any docs? If so, where?
> 2. Should we only add the view hierarchy attachment(s) in case of error or exception, just like the screenshots? Or with all events, as I have done now? [Attach UIKit view hierarchy to events #1772](https://github.com/getsentry/sentry-cocoa/issues/1772) suggests all events.

1. Yes. You need to fork https://github.com/getsentry/sentry-docs and them submit a PR. Look for screenshots, basically Is the same thing, instead of image we have a view hierarchy file.
2.  Yes, only for errors for now. Exactly like screenshots.

We still need to define the file format and what information to use. 
@bruno-garcia  thoughts?

**Comment by kevinrenskers** [comment]
The issue doesn't mention it's only for errors though, so are you sure? 

And as for file format... this comes straight from iOS itself, and contains everything you'd want to know from the view hierarchy. Creating a format ourselves would obviously be a lot more work. Not sure why you'd invent your own format?

**Comment by brustolin** [comment]
> Creating a format ourselves would obviously be a lot more work. Not sure why you'd invent your own format?

We may want to have a standard for all platforms.

**Comment by brustolin** [comment]
> The issue doesn't mention it's only for errors though, so are you sure?

Kind. We had an informal discussion on this. Now is the moment to improve the specs.

**Comment by kevinrenskers** [comment]
Docs PR: https://github.com/getsentry/sentry-docs/pull/5400

**Comment by kevinrenskers** [comment]
We could keep it as an attachment for now and in the future when the Sentry frontend can render a specialized format more nicely, we can always change it at that point. But I think this already adds a nice added value, even if we don't standardize the view hierarchy format.

File: Sources/Sentry/Public/SentryOptions.h [hunk 1]
```diff
@@ -206,6 +206,13 @@ NS_SWIFT_NAME(Options)
  */
 @property (nonatomic, assign) BOOL attachScreenshot;
 
+/**
+ * Automatically attaches a textual representation of the view hierarchy when capturing an event.
+ *
```
Comment (brustolin) [inline Sources/Sentry/Public/SentryOptions.h:211]:
```suggestion
 * This feature is EXPERIMENTAL.
 *
 * Automatically attaches a textual representation of the view hierarchy when capturing an error event.
 *
```

**Review by brustolin — APPROVED** [review]
> It just seems like a waste to hold this feature back while all platforms align this. Like I said, we can always follow up with an update.

That's a good point

**Pull request merged**